### ISSUE TITLE
Row and column of non-symmetric matrix in HiGlass heatmap

### ISSUE THREAD

**Issue by frankyan** [higlass/higlass#704]
I used cooler 0.7.10-dev version (asymmetric branch) to generate non-symmetric `mcool` file from RNA-DNA interaction `.pairs` data. HiGlass can load the non-symmetric `mcool` file as a non-symmetric heatmap.
I found that the row in the heatmap is "c2:p2" in `cool` file and the column is "c1:p1" in `cool` file. However, I think it's better to transpose the heatmap view, i.e., row is "c1:p1" and column is "c2:p2". Is there any options in HiGlass can change the view?

**Comment by pkerpedjiev** [comment]
Hey, thanks for reporting this. @nvictus and I are looking at it right now. Should hopefully have a release with a fix in the next few days.

**Comment by frankyan** [comment]
Thanks. By the way, will the cooler Schema V3 non-symmetric matrix be supported by HiGlass in the next release? Current HiGlass version cannot correctly display the non-symmetric matrix generated by cooler v0.8 because it uses new Schema V3. The problem has been reported in cooler repo too.
https://github.com/mirnylab/cooler/issues/119#issuecomment-504705273

**Comment by pkerpedjiev** [comment]
Yes. Just to confirm, @nvictus that's what this PR (https://github.com/higlass/clodius/pull/95) implements, right?

**Comment by nvictus** [comment]
yup!

**Comment by frankyan** [comment]
I saw that a new HiGlass release. I tried it. It's great that it supports asymmetric mcool file generated by cooler v0.8.5.
In RNA-DNA interaction heatmap, we always use rows of heatmap to represent RNA, and columns to represent chromatin position. We store RNA read position as "c1:p1" and DNA read position as "c2:p2" in pairs format file. For example, an RNA-DNA interaction pair of `RNA: chrX:100100-100200` and `DNA: chr1:200000-200100`, it should be plotted at the cross point of row `chrX:100100-100200` and column `chr1:200000-200100`, which is in the lower-left triangle. However, HiGlass still plot it at the mirrored point in the upper-right triangle. Could you please change the default setting to transpose current view of the asymmetric matrix? Thanks.
Alternatively, is it possible to set the orientation when we ingest data or when we visualize it in browser?

**Comment by pkerpedjiev** [comment]
Sorry, we're still working on applying patches and stabilizing the new `v1.6` release. We're looking into ways to add this functionality into one of the upcoming patches.

**Comment by frankyan** [comment]
Hi, the 4DN DCIC is implementing iMARGI RNA-DNA interaction mapping pipeline in DCIC Data Portal. It also has Higlass visualization. For visualization consistency, I'd like to know the future plan for this issue of asymmetric heatmap "orientation".
We store RNA read position as "c1:p1" and DNA read position as "c2:p2" in `pairs` format file. In the HiGlass heatmap, we hope to use rows of heatmap to represent RNA, and columns to represent DNA position (RNA x DNA matrix). However, the orientation of Higlass heatmap view of the `mcool` file directly generated from `pairs` file is reversed (DNA x RNA matrix).

Do you have a plan to change the default heatmap view orientation for asymmetric `mcool`, or add a transpose function button in Higlass control panel? 
- If you have the plan and it can be released soon, then we will directly generate `mcool` from `pairs` file in iMARGI pipeline.
- If it won't be applied, then we will generate `mcool` from reversed `pairs` file in the iMARGI pipeline.

Both solutions are OK for us. We just need to know your plan to determine our solution. Thank you.

**Comment by frankyan** [comment]
Hi Peter,

Thank you for your suggestion. The transpose option in the front end might be not necessary. I think we can solve the problem in the backend. 

In my opinion, the best solution should satisfy three requirements.

1. The output files of further operations on `.cool/mcool` file, such as dumping binned contact matrix, should consistent with the `.pairs` file, i.e., `c1:p1` is RNA and `c2:p2` is DNA.
2. Users don't need to generate two `cool/mcool` files, transposed for visualization and non-transposed for other operations, such as dumping binned contact matrix. 
3. Users can know the orientation of Higlass view.

I think adding a flag into `cool/mcool` header might a good idea. I suggest the following solution. The key points are point 3 and 4. If these two points are reachable, I think everything will be solved. 

1. Always generate `cool/mcool` file directly from `.pairs` file. Never use a reversed `.pairs` file. So the output of all the analysis operations on `cool/mcool` file are consistent with the original `.pairs` file.
2. A `higlass heatmap orientation` flag is required when setting the `asymmetric matrix` flag for generating `cool` from `pairs` file, and it only works with `asymmetric matrix` flag is true. 
3. I think HiGlass uses X-Y coordinates system when rendering heatmap, and `c1:p1` is coordinate on X-axis and `c2:p2` is coordinate on Y-axis in default. So the `higlass heatmap orientation` flag can instruct the HiGlass how to assign the two coordinates to X and Y-axis when rendering asymmetric matrix. 
4. The default value (also for already generated files without the flag) of `higlass heatmap orientation` flag is `1y2x`, so HiGlass use `c1:p1` value as coordinate of Y-axis and `c2:p2` value as the coordinate of X-axis. If set `higlass heatmap orientation` flag to `1x2y`, then HiGlass will use `c1:p1` value as the coordinate of X-axis and `c2:p2` value as the coordinate of Y-axis, which is current used orientation. 
5. Users can know the orientation of viewing in HiGlass by checking the `higlass heatmap orientation` flag in header of `cool` file. Besides, I hope there will be a way to indicate the orientation in front end when users cannot access the `cool` file.

I am sorry that I don't know the actual way of rendering heatmap in HiGlass, so the suggestion might be meaningless. What do you think? Thank you.

Best,

Frank

**Comment by nvictus** [comment]
Hi Frank,

I think we should be able to solve this almost entirely on the server side (https://github.com/higlass/clodius/pull/105), so that bin1 gets reinterpreted as rows (vertical) and bin2 as columns (horizontal). It's a trivial swap of variables in `clodius` but will require a compensatory change in the way the half matrix is mirrored on the client. 
With that change, the default should be as you expect: RNA = i = rows = vertical, DNA = j = cols = horizontal. This way we can keep the default unsurprising. 

That opens the possibility of introducing a client-side track option to opt into transposing the heatmap, or even rotating the whole view plane 90 degrees left (turning i into x and j into y).

Provided the changes on the client are trivial and we can handle backwards compatibility, I propose this be the way to go.

Best,

Nezar

**Comment by frankyan** [comment]
Hi Nezar,

That's great. Thank you. Looking forward to the update.

Best regards,

Frank

**Comment by luoxun0213** [comment]
> Hey, thanks for reporting this. @nvictus and I are looking at it right now. Should hopefully have a release with a fix in the next few days.

Hi Peter,

Recently I used HiGlass (`higlass-docker:0.10.4`) to visualize the RNA-DNA interaction heatmap, and I found it still exists that the row in the heatmap is "c2:p2" in the cool file and the column is "c1:p1" in the cool file. So are there any methods to transpose the heatmap view in HiGlass? 

Best,

Xianglei

**Comment by manzt** [comment]
>  So are there any methods to transpose the heatmap view in HiGlass?

Unfortunately there are not and we do not have this on our roadmap. Please feel free to open a separate issue with a request for this feature for further discussion!
